subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,742 @@
+			contains_query = true;
+			if (path_length > i + 1) {
+				path_start = i + 1;
+			} else {
+				path_start = path_length;
+			}
+			break;
+		}
+	}
+
+	if (contains_query) {
+		for (i = path_start; i < path_length; i++) {
+			path_end = i;
+			if (path[i] == COAP_PATH_ELEM_AMP || path[i] == COAP_PATH_ELEM_QUERY) {
+				/* Guard for preceding delimiters */
+				if (path_start < path_end) {
+					ret = coap_packet_append_option(pckt, COAP_OPTION_URI_QUERY,
+									path + path_start,
+									path_end - path_start);
+					if (ret < 0) {
+						LOG_ERR("Failed to append path to CoAP message");
+						goto out;
+					}
+				}
+				/* Check if there is a new query option after delimiter,
+				 * if not, point to the end of string to not add
+				 * new option after this
+				 */
+				if (path_length > i + 1) {
+					path_start = i + 1;
+				} else {
+					path_start = path_length;
+				}
+			}
+		}
+	}
+
+	if (path_start < path_end) {
+		if (contains_query) {
+			ret = coap_packet_append_option(pckt, COAP_OPTION_URI_QUERY,
+							path + path_start,
+							path_end - path_start + 1);
+		} else {
+			ret = coap_packet_append_option(pckt, COAP_OPTION_URI_PATH,
+							path + path_start,
+							path_end - path_start + 1);
+		}
+		if (ret < 0) {
+			LOG_ERR("Failed to append path to CoAP message");
+			goto out;
+		}
+	}
+
+out:
+	return ret;
+}
+
+static enum coap_block_size coap_client_default_block_size(void)
+{
+	switch (CONFIG_COAP_CLIENT_BLOCK_SIZE) {
+	case 16:
+		return COAP_BLOCK_16;
+	case 32:
+		return COAP_BLOCK_32;
+	case 64:
+		return COAP_BLOCK_64;
+	case 128:
+		return COAP_BLOCK_128;
+	case 256:
+		return COAP_BLOCK_256;
+	case 512:
+		return COAP_BLOCK_512;
+	case 1024:
+		return COAP_BLOCK_1024;
+	}
+
+	return COAP_BLOCK_256;
+}
+
+static int coap_client_init_request(struct coap_client *client,
+				    struct coap_client_request *req)
+{
+	int ret = 0;
+	int i;
+
+	memset(client->send_buf, 0, sizeof(client->send_buf));
+	ret = coap_packet_init(&client->request, client->send_buf, MAX_COAP_MSG_LEN, 1,
+			       req->confirmable ? COAP_TYPE_CON : COAP_TYPE_NON_CON,
+			       COAP_TOKEN_MAX_LEN, coap_next_token(), req->method,
+			       coap_next_id());
+
+	if (ret < 0) {
+		LOG_ERR("Failed to init CoAP message %d", ret);
+		goto out;
+	}
+
+	ret = coap_client_init_path_options(&client->request, req->path);
+
+	if (ret < 0) {
+		LOG_ERR("Failed to parse path to options %d", ret);
+		goto out;
+	}
+
+	ret = coap_append_option_int(&client->request, COAP_OPTION_CONTENT_FORMAT, req->fmt);
+
+	if (ret < 0) {
+		LOG_ERR("Failed to append content format option");
+		goto out;
+	}
+
+	/* Blockwise receive ongoing, request next block. */
+	if (client->recv_blk_ctx.current > 0) {
+		ret = coap_append_block2_option(&client->request, &client->recv_blk_ctx);
+
+		if (ret < 0) {
+			LOG_ERR("Failed to append block 2 option");
+			goto out;
+		}
+	}
+
+	/* Add extra options if any */
+	for (i = 0; i < req->num_options; i++) {
```

Comment:
This is where I meant to comment about `i`, but I can see it defined above now on line 196.

When I made the comment earlier, I received a compile error on your code until I defined `i`.  Strange.